src/deploy/osp_deployer/powerflexgw.py

```diff
@@ -0,0 +1,257 @@
+              " rpm -ivh " + \
+              self.rpm_dir + \
+              "/" + \
+              self.settings.powerflex_gateway_rpm
+        self.run_as_root(cmd)
+
+
+    def configure_gateway(self):
+
+        logger.debug("Retrieving MDM IPs")
+
+        mdm_ips = []
+        properties = '/opt/emc/scaleio/gateway/webapps/ROOT/WEB-INF/classes/gatewayUser.properties'
+        director = Director()
+
+        stor_ = self.settings.storage_network.rsplit(".", 1)[0]
+        stor_.replace(".", '\.')
+
+        re = director.run_as_root("cat /var/lib/mistral/" +
+                                  self.settings.overcloud_name + 
+                                  "/powerflex-ansible/inventory.yml " +
+                                  "| sed -n '/mdms/,/tbs/{//!p}'")
+        mdm_nodes = re[0].split("\n")
+        mdm_nodes.pop()
+
+        for each in mdm_nodes:
+            provisioning_ip = each.split(" ")[0]
+            ssh_opts = (
+                " -o StrictHostKeyChecking=no "
+                " -o UserKnownHostsFile=/dev/null "
+                " -o KbdInteractiveDevices=no")
+
+            cmd = ("ssh " + ssh_opts + " heat-admin@" +
+                   provisioning_ip +
+                   " /sbin/ifconfig | grep inet.*" +
+                   stor_ +
+                   " | awk '{print $2}'")
+            re = director.run_tty(cmd)
+
+            mdm_ip = re[0].split("\n")[1]
+            mdm_ips.append(mdm_ip.strip())
+
+        logger.debug("Updating gatewayUser.properties")
+
+        cmd = ('sed -i "s|^mdm.ip.addresses=*|' +
+               'mdm.ip.addresses=' +
+               mdm_ips[0] +
+               ',' +
+               mdm_ips[1] +
+               '|" ' +
+               properties)
+
+        self.run_as_root(cmd)
+
+
+    def get_ssl_certificates(self):
+
+        cmd = "mkdir -p " + self.certs_dir
+        self.run_as_root(cmd)
+
+        logger.debug("Retrieving all nodes IPs")
+
+        node_ips = []
+        director = Director()
+
+        cmd = "grep " + self.ip + " ~./ssh/known_hosts"
+        re = director.run(cmd)
+        
+        if self.ip not in re:
+            logger.debug("First time connection to the powerflex gateway vm")
+            cmd = "ssh-keyscan -t ecdsa " + self.ip + " >> ~/.ssh/known_hosts"
+            re = director.run(cmd)
+
+        ssh_opts = (
+        " -o StrictHostKeyChecking=no "
+        " -o UserKnownHostsFile=/dev/null "
+        " -o KbdInteractiveDevices=no"
+        " -o LogLevel=ERROR")
+
+        re = director.run(director.source_stackrc +
+                              "openstack server list -f value")
+        nodes = re[0].split("\n")
+        nodes.pop()
+
+        for each in nodes:
+            hostname = each.split(" ")[1]
+            ip = each.split(" ")[3].split("=")[1]
+
+            cmd = ("ssh " + ssh_opts + " heat-admin@" +
+                   ip + 
+                   " test -f " + self.mdm_cert + "; echo $?;")
+                 
```

Comment:
Dont we know what node types the mdm's would be running on ? Could loop through those types of nodes only, rather than go through all of the overcloud nodes

Reply:
The way the services are deployed is based upon the service layout that the user can modify and custom. If we don't want to go trough each of the nodes, we can take the ansible inventory as a reference. That way, we can determine safely on what node the mdm is running, which are generally the contollers.

Reply:
Will MDM not _always_ be running on controllers ?

Reply:
not necessarily. As I said, it depends on the service layout the user wans to apply. We need to keep it flexible, this is the way the Ansible playbook are designed. Better not to hardcode the nodes the MDMs are running on.

Reply:
Not scanning all the hosts anymore. The MDM lookup is based upon the ansible inventory file.